feflow/protocols/nonequilibrium_cycling.py

```diff
@@ -165,26 +166,32 @@ def _execute(self, ctx, *, state_a, state_b, mapping, settings, **inputs):
         # Note: by default this is cached to ctx.shared/db.json so shouldn't
         # incur too large a cost
         self.logger.info("Parameterizing molecules")
-        small_mols_a = []
-        for comp in state_a.components.values():
-            if isinstance(comp, SmallMoleculeComponent):
-                small_mols_a.append(comp)
-
-        for comp in small_mols_a:
-            offmol = comp.to_openff()
-            system_generator.create_system(offmol.to_topology().to_openmm(),
-                                           molecules=[offmol])
-            if comp == ligand_a:
-                mol_b = ligand_b.to_openff()
-                system_generator.create_system(mol_b.to_topology().to_openmm(),
-                                               molecules=[mol_b])
+        # TODO: Refactor if/when gufe provides the functionality https://github.com/OpenFreeEnergy/gufe/issues/251
+        # The following creates a dictionary with all the small molecules in the states, with the structure:
+        #    Dict[SmallMoleculeComponent, openff.toolkit.Molecule]
+        state_a_small_mols = {component: component.to_openff() for component in state_a.components.values() if
+                              isinstance(component, SmallMoleculeComponent)}
+        state_b_small_mols = {component: component.to_openff() for component in state_b.components.values() if
+                              isinstance(component, SmallMoleculeComponent)}
+
+        # Assign charges if unassigned -- more info: Openfe issue #576
+        for off_mol in chain(state_a_small_mols.values(), state_b_small_mols.values()):
```

Comment:
```suggestion
        combined_small_mols = state_a_small_mols | state_b_small_mols

        # Assign charges if unassigned -- more info: Openfe issue #576
        for off_mol in combined_small_mols.values():
```

Reply:
I believe doing the above should avoid the duplication of charging @IAlibay is referring to [here](https://github.com/choderalab/feflow/pull/13/files#r1408469035).

Reply:
We're relying on the pre-existance of partial charges as the way to ensure that we then don't re-generate partial charges later.

Depending on how OFF molecule equality works, wouldn't doing this end up with some molecules without partial charges in some of the two object lists?